## **Tugas Besar 2 Pengenalan Komputasi**
## **Bagian 5**
### **Kelompok 6**
- Ester Tri Wahyuningsih (16022101)
- Rayyan Hannata Denan   (16022106)
- Valentinus Brian Johan (16022111)
- Brian Matthew          (16022116)

### Import modul-modul yang dibutuhkan dari library

In [14]:
import pandas as pd
from matplotlib import pyplot as plt

### File 'supermarket_sales - Sheet1.csv' dimasukkan ke variabel bernama df menggunakan library pandas

In [15]:
df = pd.read_csv('supermarket_sales - Sheet1.csv')

### Menampilkan dataframe

In [16]:
df

Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
1    226-31-3081      C  Naypyitaw        Normal  Female   
2    631-41-3108      A     Yangon        Normal    Male   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
..           ...    ...        ...           ...     ...   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   
997  727-02-1313      A     Yangon        Member    Male   
998  347-56-2442      A     Yangon        Normal    Male   
999  849-09-3807      A     Yangon        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle       46.33         7  16.2155   340.5255   
3         Health and beauty       58.22         8  23.2880   489.0480   
4         Sports and travel       86.31         7  30.2085   634.3785   
..                      ...         ...       ...      ...        ...   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   
997      Food and beverages       31.84         1   1.5920    33.4320   
998      Home and lifestyle       65.82         1   3.2910    69.1110   
999     Fashion accessories       88.34         7  30.9190   649.2990   

          Date   Time      Payment    cogs  gross margin percentage  \
0     1/5/2019  13:08      Ewallet  522.83                 4.761905   
1     3/8/2019  10:29         Cash   76.40                 4.761905   
2     3/3/2019  13:23  Credit card  324.31                 4.761905   
3    1/27/2019  20:33      Ewallet  465.76                 4.761905   
4     2/8/2019  10:37      Ewallet  604.17                 4.761905   
..         ...    ...          ...     ...                      ...   
995  1/29/2019  13:46      Ewallet   40.35                 4.761905   
996   3/2/2019  17:16      Ewallet  973.80                 4.761905   
997   2/9/2019  13:22         Cash   31.84                 4.761905   
998  2/22/2019  15:33         Cash   65.82                 4.761905   
999  2/18/2019  13:28         Cash  618.38                 4.761905   

     gross income  Rating  
0         26.1415     9.1  
1          3.8200     9.6  
2         16.2155     7.4  
3         23.2880     8.4  
4         30.2085     5.3  
..            ...     ...  
995        2.0175     6.2  
996       48.6900     4.4  
997        1.5920     7.7  
998        3.2910     4.1  
999       30.9190     6.6  

[1000 rows x 17 columns]

# **Sampel Data**

### **Sampel Data 1**

In [17]:
df[:1]

Invoice ID Branch    City Customer type  Gender       Product line  \
0  750-67-8428      A  Yangon        Member  Female  Health and beauty   

   Unit price  Quantity   Tax 5%     Total      Date   Time  Payment    cogs  \
0       74.69         7  26.1415  548.9715  1/5/2019  13:08  Ewallet  522.83   

   gross margin percentage  gross income  Rating  
0                 4.761905       26.1415     9.1

#### Dapat dilihat bahwa untuk suatu baris data diberikan informasi berupa *Invoice ID*, *Branch*, *City*, *Customer type*, *Gender*, *Product line*, *Unit price*, *Quantity*, *Tax 5%*, *Total*, *Date*, *Time*, *Payment*, *cogs*, *gross margin percentage*, *gross income*, dan *rating*.

### **Sampel Data 2**

In [18]:
max_U = df['Unit price'].max()
df.loc[df['Unit price'] == max_U]

Invoice ID Branch       City Customer type Gender       Product line  \
122  219-22-9386      B   Mandalay        Member   Male  Sports and travel   
983  148-41-7930      C  Naypyitaw        Normal   Male  Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   

       cogs  gross margin percentage  gross income  Rating  
122  899.64                 4.761905        44.982     4.2  
983  699.72                 4.761905        34.986     6.1

#### Sampel data diatas memuat data mengenai transaksi dengan *unit price* tertinggi yaitu 99.96.
#### Bisa dilihat juga bahwa terdapat 2 data dengan *unit price* 99.96.
#### Satu dari produk *sports and travel* dan satu lagi dari produk *health and beauty*. 

### **Sampel Data 3**

In [19]:
id = df['Invoice ID'].str.contains('101-', case=False, na = False )
df[id]

Invoice ID Branch       City Customer type  Gender        Product line  \
162  101-17-6199      A     Yangon        Normal    Male  Food and beverages   
867  101-81-4070      C  Naypyitaw        Member  Female   Health and beauty   

     Unit price  Quantity   Tax 5%     Total       Date   Time      Payment  \
162       45.79         7  16.0265  336.5565  3/13/2019  19:44  Credit card   
867       62.82         2   6.2820  131.9220  1/17/2019  12:36      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
162  320.53                 4.761905       16.0265     7.0  
867  125.64                 4.761905        6.2820     4.9

In [20]:
id2 = df['Invoice ID'].str.contains('156-', case=False, na = False )
df[id2]

Invoice ID Branch      City Customer type  Gender  \
759  156-95-3964      A    Yangon        Normal  Female   
978  156-20-0370      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
759      Food and beverages       55.39         4  11.0780  232.6380   
978  Electronic accessories       25.45         1   1.2725   26.7225   

          Date   Time      Payment    cogs  gross margin percentage  \
759  3/25/2019  15:19      Ewallet  221.56                 4.761905   
978  3/10/2019  18:10  Credit card   25.45                 4.761905   

     gross income  Rating  
759       11.0780     8.0  
978        1.2725     5.1

In [21]:
id3 = df['Invoice ID'].str.contains('123-', case=False, na = False )
df[id3]

Invoice ID Branch       City Customer type  Gender       Product line  \
3    123-19-1176      A     Yangon        Member    Male  Health and beauty   
597  123-35-4896      C  Naypyitaw        Normal  Female  Sports and travel   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment    cogs  \
3         58.22         8  23.288  489.048  1/27/2019  20:33  Ewallet  465.76   
597       46.66         9  20.997  440.937  2/17/2019  19:11  Ewallet  419.94   

     gross margin percentage  gross income  Rating  
3                   4.761905        23.288     8.4  
597                 4.761905        20.997     5.3

#### 3 sampel data diatas memuat transaksi yang memiliki invoice ID dengan 3 angka pertama yang sama yaitu 101, 156, dan 123. 
#### Bisa dilihat bahwa tidak ada kesamaan yang konsisten pada kolom lain untuk setiap 3 angka yang sama. 
#### Hal itu berarti invoice ID merupakan bilangan acak yang tidak mengikuti pola tertentu.

### **Sampel Data 4**

In [22]:
df.loc[:5,['Invoice ID', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'cogs', 'gross income']]

Invoice ID  Unit price  Quantity   Tax 5%     Total    cogs  gross income
0  750-67-8428       74.69         7  26.1415  548.9715  522.83       26.1415
1  226-31-3081       15.28         5   3.8200   80.2200   76.40        3.8200
2  631-41-3108       46.33         7  16.2155  340.5255  324.31       16.2155
3  123-19-1176       58.22         8  23.2880  489.0480  465.76       23.2880
4  373-73-7910       86.31         7  30.2085  634.3785  604.17       30.2085
5  699-14-3026       85.39         7  29.8865  627.6165  597.73       29.8865

#### Sampel data diatas memuat aspek-aspek keuangan dari transaksi
#### Bila diperhatikan, kolom 'Total' mengikuti rumus : 
$ 
Total = [Unit price * Quantity] + \frac{5}{100}* [Unit price * Quantity] 
$ 
#### Dan kolom 'gross income' mengikuti rumus :
$
grossincome = Total - cogs 
$

### **Sampel data 5**

In [23]:
pd.DataFrame(df['Date'].sort_values())

Date
496  1/1/2019
484  1/1/2019
567  1/1/2019
523  1/1/2019
970  1/1/2019
..        ...
393  3/9/2019
73   3/9/2019
433  3/9/2019
585  3/9/2019
84   3/9/2019

[1000 rows x 1 columns]

In [24]:
pd.DataFrame(df['Time'].sort_values())

Time
913  10:00
180  10:00
293  10:00
200  10:01
669  10:01
..     ...
106  20:54
267  20:55
893  20:57
136  20:59
935  20:59

[1000 rows x 1 columns]

#### Sampel data tersebut memuat informasi mengenai waktu.

#### Dari kolom 'Date' yang telah diurutkan, bisa disimpulkan bahwa data diambil dari 1 Januari 2019 sampai 30 Maret 2019.

#### Dari kolom 'Time' yang telah diurutkan, bisa disimpulkan bahwa toko buka pukul 10 pagi dan tutup pukul 9 malam.

# **Statistik**

## **Rata-Rata**

### Memuat informasi mengenai nilai harapan dari suatu atribut data

In [25]:
# menggunakan fungsi
df[['atribut data 1', 'atribut data 2', ..., 'atribut data n']].mean()

KeyError: "None of [Index(['atribut data 1', 'atribut data 2', Ellipsis, 'atribut data n'], dtype='object')] are in the [columns]"

In [ ]:
mean_c = pd.DataFrame(df[['Unit price', 'Quantity', 'Tax 5%', 'Total', 'gross margin percentage', 'gross income', 'Rating']].mean())
mean_c.rename(columns = {0: 'Rata-rata'}, inplace = True)
mean_c

Rata-rata
Unit price                55.672130
Quantity                   5.510000
Tax 5%                    15.379369
Total                    322.966749
gross margin percentage    4.761905
gross income              15.379369
Rating                     6.972700

## **Standar Deviasi**

### Memuat informasi mengenai seberapa bervariasi dan tersebar nilai-nilai dari suatu atribut data. Semakin kecil standar deviasi maka semakin dekat nilainya dengan nilai rata-rata (tidak bervariasi/homogen), sebaliknya semaking jauh maka semakin jauh nilainya dengan rata-rata (bervariasi)

In [ ]:
# menggunakan fungsi
df[['atribut data 1', 'atribut data 2', ..., 'atribut data n']].std()

In [ ]:
std_c = pd.DataFrame(df[['Unit price', 'Quantity', 'Tax 5%', 'Total', 'gross margin percentage', 'gross income', 'Rating']].std())
std_c.rename(columns = {0: 'Standar deviasi'}, inplace = True)
std_c

Standar deviasi
Unit price                     26.494628
Quantity                        2.923431
Tax 5%                         11.708825
Total                         245.885335
gross margin percentage         0.000000
gross income                   11.708825
Rating                          1.718580

## **Nilai Maksimum**

### Memuat informasi nilai terbesar suatu atribut data

In [ ]:
# menggunakan fungsi
df[['atribut data 1', 'atribut data 2', ..., 'atribut data n']].max()

In [ ]:
max_c = pd.DataFrame(df[['Unit price', 'Quantity', 'Tax 5%', 'Total', 'gross margin percentage', 'gross income', 'Rating']].max())
max_c.rename(columns = {0: 'Nilai maksimum'}, inplace = True)
max_c

Nilai maksimum
Unit price                    99.960000
Quantity                      10.000000
Tax 5%                        49.650000
Total                       1042.650000
gross margin percentage        4.761905
gross income                  49.650000
Rating                        10.000000

## **Nilai Minimum**

### Memuat informasi nilai terkecil suatu atribut data

In [ ]:
# menggunakan fungsi
df[['atribut data 1', 'atribut data 2', ..., 'atribut data n']].min()

In [ ]:
min_c = pd.DataFrame(df[['Unit price', 'Quantity', 'Tax 5%', 'Total', 'gross margin percentage', 'gross income', 'Rating']].min())
min_c.rename(columns = { 0: 'Nilai minimum'}, inplace = True)
min_c

Nilai minimum
Unit price                   10.080000
Quantity                      1.000000
Tax 5%                        0.508500
Total                        10.678500
gross margin percentage       4.761905
gross income                  0.508500
Rating                        4.000000

## **Distribusi Frekuensi**

### Memuat informasi mengenai frekuensi suatu atribut data di seluruh sebaran data

In [ ]:
# kebanyakan menggunakan fungsi
df['kolom'].value_counts
# atau jika data merupakan interval [a,b], memanfaatkan fungsi
len(df.loc[(df['atribut data'] > intervalA ) & (df['atribut data'] < intervalB) ])

### Distribusi frekuensi kolom 'Branch' (Cabang tempat transaksi dilakukan)

In [ ]:
B = pd.DataFrame(df['Branch'].value_counts())
B.rename(columns = {'Branch': 'Frequency'}, inplace = True)
B


Frequency
A        340
B        332
C        328

### Distribusi frekuensi kolom 'City' (Kota tempat transaksi dilakukan)

In [ ]:
C = pd.DataFrame(df['City'].value_counts())
C.rename(columns = {'City': 'Frequency'}, inplace = True)
C


Frequency
Yangon           340
Mandalay         332
Naypyitaw        328

### Distribusi frekuensi kolom 'Customer type' (Jenis membership pelanggan)

In [ ]:
CT = pd.DataFrame(df['Customer type'].value_counts())
CT.rename(columns = {'Customer type': 'Frequency'}, inplace = True)
CT


Frequency
Member        501
Normal        499

### Distribusi frekuensi kolom 'Gender' (Jenis kelamin pelanggan)

In [ ]:
G = pd.DataFrame(df['Gender'].value_counts())
G.rename(columns = {'Gender': 'Frequency'}, inplace = True)
G

Frequency
Female        501
Male          499

### Distribusi frekuensi kolom 'Product line' (Jenis barang yang dibeli)

In [ ]:
PL = pd.DataFrame(df['Product line'].value_counts())
PL.rename(columns = {'Product line': 'Frequency'}, inplace = True)
PL

Frequency
Fashion accessories           178
Food and beverages            174
Electronic accessories        170
Sports and travel             166
Home and lifestyle            160
Health and beauty             152

### Distribusi frekuensi kolom 'Quantity' (Jumlah barang yang dibeli)

In [ ]:
Q = pd.DataFrame(df['Quantity'].value_counts())
Q.rename(columns = {'Quantity': 'Frequency'}, inplace = True)
Q

Frequency
10        119
1         112
4         109
7         102
5         102
6          98
9          92
2          91
3          90
8          85

### Distribusi frekuensi kolom 'Date' dibagi berdasarkan bulan

In [ ]:
df['Date'] = pd.to_datetime(df['Date'])
df2 = df.groupby(df['Date'].dt.month).size().reset_index()
df2.rename(columns = {'Date': 'Month',0 : 'Frequency'}, inplace = True)
df2.replace([1,2,3], ['January', 'February', 'March'])

Month  Frequency
0   January        352
1  February        303
2     March        345

### Distribusi frekuensi kolom Time dibagi tiap range waktu sebesar 2 jam

In [ ]:
Tot_1012 = len(df.loc[(df['Time'] >= '10') & (df['Time'] < '12' )])
Tot_1214 = len(df.loc[(df['Time'] >= '12') & (df['Time'] < '14' )])
Tot_1416 = len(df.loc[(df['Time'] >= '14') & (df['Time'] < '16' )])
Tot_1618 = len(df.loc[(df['Time'] >= '16') & (df['Time'] < '18' )])
Tot_1820 = len(df.loc[(df['Time'] >= '18') & (df['Time'] <= '20' )])
Tot_20top = len(df.loc[df['Time'] > '20'])
T_RWaktu = {}
T_RWaktu['Time range'] = ['10.00-12.00', '12.01-14.00', '14.01-16.00', '16.01-18.00', '18.01-20.00', '20.01-20.59']
T_RWaktu['Frequency'] = [Tot_1012, Tot_1214, Tot_1416, Tot_1618, Tot_1820, Tot_20top]
dfw = pd.DataFrame(data=T_RWaktu)
dfw

Time range  Frequency
0  10.00-12.00        191
1  12.01-14.00        192
2  14.01-16.00        185
3  16.01-18.00        151
4  18.01-20.00        206
5  20.01-20.59         75

### Distribusi frekuensi kolom 'Payment' (metode pembayaran)

In [ ]:
pd.DataFrame(df['Payment'].value_counts())

Payment
Ewallet          345
Cash             344
Credit card      311

### Distribusi frekuensi kolom 'rating'

In [ ]:
rat1 = len(df.loc[df['Rating'] < 2.0])
rat2 = len(df.loc[(df['Rating'] >= 2.0) & (df['Rating'] < 4.0)])
rat3 = len(df.loc[(df['Rating'] >= 4.0) & (df['Rating'] < 6.0)])
rat4 = len(df.loc[(df['Rating'] >= 6.0) & (df['Rating'] < 8.0)])
rat5 = len(df.loc[(df['Rating'] >= 8.0) & (df['Rating'] <= 10.0)])
T_R = {}
T_R['Rating'] = ['1-2', '2-4', '4-6', '6-8', '8-10']
T_R['Frekuensi'] = [rat1, rat2, rat3, rat4, rat5]
dfr = pd.DataFrame(data=T_R)
dfr

Rating  Frekuensi
0    1-2          0
1    2-4          0
2    4-6        315
3    6-8        356
4   8-10        329

## **Persentil**

### memuat informasi mengenai Sekumpulan nilai yang sudah diurutkan dari yang terkecil hingga terbesar lalu dibagi menjadi beberapa bagian

In [ ]:
# menggunakan fungsi
df['atribut data'].quantile(q = [a, b , c , d, ... ,z])

### Persentil kolom 'Unit price'

In [ ]:
pd.DataFrame(df['Unit price'].quantile(q=[0.1, 0.25, 0.5, 0.75, 0.9]))

Unit price
0.10      19.313
0.25      32.875
0.50      55.230
0.75      77.935
0.90      93.122

### Persentil kolom 'Quantity'

In [ ]:
pd.DataFrame(df['Quantity'].quantile(q=[0.1, 0.25, 0.5, 0.75, 0.9]))

Quantity
0.10       1.0
0.25       3.0
0.50       5.0
0.75       8.0
0.90      10.0

### Persentil kolom 'Tax 5%'

In [ ]:
pd.DataFrame(df['Tax 5%'].quantile(q=[0.1, 0.25, 0.5, 0.75, 0.9]))


Tax 5%
0.10   3.243000
0.25   5.924875
0.50  12.088000
0.75  22.445250
0.90  34.233850

### Persentil kolom 'Total'

In [ ]:
pd.DataFrame(df['Total'].quantile(q=[0.1, 0.25, 0.5, 0.75, 0.9]))

Total
0.10   68.103000
0.25  124.422375
0.50  253.848000
0.75  471.350250
0.90  718.910850

### Persentil kolom 'cogs' (cost of goods sold)

In [ ]:
pd.DataFrame(df['cogs'].quantile(q=[0.1, 0.25, 0.5, 0.75, 0.9]))

cogs
0.10   64.8600
0.25  118.4975
0.50  241.7600
0.75  448.9050
0.90  684.6770

### Persentil kolom 'Gross income' 

In [ ]:
pd.DataFrame(df['gross income'].quantile(q=[0.1, 0.25, 0.5, 0.75, 0.9]))

gross income
0.10      3.243000
0.25      5.924875
0.50     12.088000
0.75     22.445250
0.90     34.233850

### Persentil kolom 'Rating'

In [ ]:
pd.DataFrame(df['Rating'].quantile(q=[0.1, 0.25, 0.5, 0.75, 0.9]))

Rating
0.10     4.5
0.25     5.5
0.50     7.0
0.75     8.5
0.90     9.4